File: python/deepstrike/collaboration/handoff.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from deepstrike.collaboration.contract import ContractCheckResult, VerificationContract

@dataclass
class HandoffArtifact:
    """
    Structured state passed between sprints / agent instances.

    Every handoff path (context renewal and sub-agent completion)
    produces a HandoffArtifact so the next agent knows not only *what was done*
    but *what has been proven*.
    """
    goal: str
    sprint: int
    progress_summary: str
    open_tasks: list[str] = field(default_factory=list)
    contract_status: list[ContractCheckResult] = field(default_factory=list)
    drift_rate_24h: float = 0.0
    blocked_on: list[str] = field(default_factory=list)


@dataclass
class ContractOutcomeInput:
    contract: VerificationContract
    check_results: list[ContractCheckResult]
    artifact: str
    success: bool
    blocked_on: Optional[list[str]] = None


class HandoffBus:
    """
    Canonical factory for HandoffArtifact.

    Every transition between agent contexts goes through one of these static methods,
    ensuring that the resulting artifact always carries contract_status.
    """

    @staticmethod
    def from_contract_outcome(inp: ContractOutcomeInput) -> HandoffArtifact:
        failed_required = [
            r for r in inp.check_results
            if not r.passed
            and any(c.id == r.criterion_id and c.required for c in inp.contract.acceptance)
        ]
        total = len(inp.check_results)
        drift = len([r for r in inp.check_results if not r.passed]) / total if total > 0 else 0.0

        if inp.success:
            summary = f"Completed: {inp.artifact[:200]}{'…' if len(inp.artifact) > 200 else ''}"
            open_tasks: list[str] = []
        else:
            summary = (
                f"Incomplete after max attempts. "
                f"{len(failed_required)} required criteria failed."
            )
            open_tasks = [f"Fix criterion: {r.criterion_id}" for r in failed_required]

        return HandoffArtifact(
            goal=inp.contract.goal,
            sprint=1,
            progress_summary=summary,
            open_tasks=open_tasks,
            contract_status=inp.check_results,
            drift_rate_24h=drift,
            blocked_on=inp.blocked_on or [],
        )

    @staticmethod
    def from_sub_agent_result(
        *,
        goal: str,
        final_message: str,
        sprint: int = 1,
    ) -> HandoffArtifact:
        return HandoffArtifact(
            goal=goal,
            sprint=sprint,
            progress_summary=final_message[:500],
        )

    @staticmethod
    def to_context_note(artifact: HandoffArtifact) -> str:
        """Render the artifact as a compact string for injection into working context."""
        lines = [
            f"[Handoff from sprint {artifact.sprint}]",
            f"Goal: {artifact.goal}",
            f"Progress: {artifact.progress_summary}",
        ]
        if artifact.open_tasks:
            lines.append(f"Open tasks: {'; '.join(artifact.open_tasks)}")
        if artifact.contract_status:
            passed = sum(1 for r in artifact.contract_status if r.passed)
            lines.append(f"Contract: {passed}/{len(artifact.contract_status)} criteria passed")
        if artifact.blocked_on:
            lines.append(f"BLOCKED ON: {'; '.join(artifact.blocked_on)}")
        if artifact.drift_rate_24h > 0:
            lines.append(f"Drift rate: {artifact.drift_rate_24h * 100:.1f}%")
        return "\n".join(lines)

    @staticmethod
    def requires_escalation(artifact: HandoffArtifact, *, drift_threshold: float = 0.05) -> bool:
        """True when drift rate exceeds threshold or blocked_on is non-empty."""
        return artifact.drift_rate_24h > drift_threshold or bool(artifact.blocked_on)

File: node/src/collaboration/handoff.ts
import type { ContractCheckResult, VerificationContract } from "./contract.js"

/**
 * HandoffArtifact — the single exchange token between sprints and agent instances.
 *
 * All handoff paths converge here:
 *   - CreatorVerifier AttemptLoop completion → HandoffBus.fromContractOutcome()
 *   - Sub-agent completion              → HandoffBus.fromSubAgentResult()
 *   - Context renewal (kernel)          → carried in kernel's HandoffArtifact type
 *
 * The invariant: a HandoffArtifact tells the next agent not only *what was done*
 * but *what has been proven*. The contract_status field is never discarded on renewal.
 */
export interface HandoffArtifact {
  goal: string
  sprint: number
  progressSummary: string
  openTasks: string[]
  /** Per-criterion verification results from the most recent contract run. */
  contractStatus: ContractCheckResult[]
  /** Ratio of verification failures over 24 h (failed / total). 0.0 if no data. */
  driftRate24h: number
  /** Issues blocking completion — require human or orchestrator attention. */
  blockedOn: string[]
}

/** Input to HandoffBus.fromContractOutcome */
export interface ContractOutcomeInput {
  contract: VerificationContract
  checkResults: ContractCheckResult[]
  artifact: string
  success: boolean
  blockedOn?: string[]
}

/**
 * HandoffBus — the canonical factory for HandoffArtifact.
 *
 * Every transition between agent contexts goes through one of these static methods.
 * This ensures that the resulting artifact always carries contract_status, never just
 * a prose summary.
 */
export class HandoffBus {
  /**
   * Build a HandoffArtifact from a creator-verifier AttemptLoop outcome.
   * The artifact field is used as the progress summary.
   */
  static fromContractOutcome(input: ContractOutcomeInput): HandoffArtifact {
    const failedRequired = input.checkResults.filter(r => {
      if (r.passed) return false
      const c = input.contract.acceptance.find(a => a.id === r.criterionId)
      return c?.required ?? false
    })

    return {
      goal: input.contract.goal,
      sprint: 1,
      progressSummary: input.success
        ? `Completed: ${input.artifact.slice(0, 200)}${input.artifact.length > 200 ? "…" : ""}`
        : `Incomplete after max attempts. ${failedRequired.length} required criteria failed.`,
      openTasks: input.success ? [] : failedRequired.map(r => `Fix criterion: ${r.criterionId}`),
      contractStatus: input.checkResults,
      driftRate24h: input.checkResults.length > 0
        ? input.checkResults.filter(r => !r.passed).length / input.checkResults.length
        : 0,
      blockedOn: input.blockedOn ?? [],
    }
  }

  /**
   * Build a HandoffArtifact from a sub-agent's final message.
   * Matches the convention established by `report_sub_agent` (only final message injected).
   */
  static fromSubAgentResult(opts: {
    goal: string
    finalMessage: string
    sprint?: number
  }): HandoffArtifact {
    return {
      goal: opts.goal,
      sprint: opts.sprint ?? 1,
      progressSummary: opts.finalMessage.slice(0, 500),
      openTasks: [],
      contractStatus: [],
      driftRate24h: 0,
      blockedOn: [],
    }
  }

  /**
   * Render the artifact as a compact injection string for the next agent's
   * working partition (not system — this is a handoff note, not a permanent rule).
   */
  static toContextNote(artifact: HandoffArtifact): string {
    const lines = [
      `[Handoff from sprint ${artifact.sprint}]`,
      `Goal: ${artifact.goal}`,
      `Progress: ${artifact.progressSummary}`,
    ]

    if (artifact.openTasks.length > 0) {
      lines.push(`Open tasks: ${artifact.openTasks.join("; ")}`)
    }

    if (artifact.contractStatus.length > 0) {
      const passed = artifact.contractStatus.filter(r => r.passed).length
      lines.push(`Contract: ${passed}/${artifact.contractStatus.length} criteria passed`)
    }

    if (artifact.blockedOn.length > 0) {
      lines.push(`BLOCKED ON: ${artifact.blockedOn.join("; ")}`)
    }

    if (artifact.driftRate24h > 0) {
      lines.push(`Drift rate: ${(artifact.driftRate24h * 100).toFixed(1)}%`)
    }

    return lines.join("\n")
  }

  /**
   * True when drift rate exceeds threshold or required criteria are blocked.
   * Use to decide whether to pause autonomous delegation and escalate.
   */
  static requiresEscalation(
    artifact: HandoffArtifact,
    opts: { driftThreshold?: number } = {},
  ): boolean {
    const threshold = opts.driftThreshold ?? 0.05
    return artifact.driftRate24h > threshold || artifact.blockedOn.length > 0
  }
}
